### PR TITLE
fix D415 - EndsInPunctuation

### PR DESCRIPTION
**Pull request by nstarman**
I also cleaned up the per-file ignore config for examples
In future, ruff will try to autofix the docstring, ensuring the summary ends in a period.

Though ruff can generally auto fix D415, for multi-line 1-line summaries and summary lines ending in a list, it doesn't know how to fix this, so I did it manually. There were around 30 such occurrences.

This PR only fixes D415 and not additional pydocstyle errors because 165 files are affected and some fixes required manual intervention and therefore more detailed scrutiny by maintainers.

Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Review by eerovaher — CHANGES_REQUESTED** [review]
The function docstrings in `astropy/coordinates/angle_formats.py` and `astropy/units/format/*.py` are regular expressions, so adding a period changes them.

**Review by mhvk — CHANGES_REQUESTED** [review]
Like @eerovaher noted for coordinates, also for units the parser/lexer docstrings are regex's and should not get final periods. That also means we should not set auto-fix to this (unless it can be fixed somehow, e.g., by not doing it on raw strings).

**Comment by nstarman** [comment]
> The function docstrings in `astropy/coordinates/angle_formats.py` and `astropy/units/format/*.py` are regular expressions, so adding a period changes them.

Thanks for the catch. I'll set ruff to ignore D400 on those files.

**Comment by nstarman** [comment]
I've added ignores for

```
"astropy/coordinates/angle_formats.py" = ["D400"]
"astropy/coordinates/angle_lextab.py" = ["D400"]
"astropy/coordinates/angle_parsetab.py" = ["D400"]
"astropy/units/format/*.py" = ["D400"]
```

File: astropy/units/core.py
Comment (mhvk) [inline astropy/units/core.py:86]: resultant -> resulting

File: astropy/units/core.py
Comment (mhvk) [inline astropy/units/core.py:775]: Resultant -> Resulting

File: astropy/utils/data.py
Comment (mhvk) [inline astropy/utils/data.py:172]: Oops, it really shouldn't add a dot after a `?` (or `!`, not sure we have those!)

Bug in ruff?

**Review by mhvk — COMMENTED** [review]
Thanks! units, utils and time are OK

File: astropy/utils/data.py
Comment (nstarman) [reply astropy/utils/data.py:172]: I've opened https://github.com/charliermarsh/ruff/issues/1858.
D401 says the line should be in an imperative tense, so even though that's not (yet) turned on, I've rephrased this one-line docstring to "Whether the `url` requires the optional ``fsspec`` dependency to open."

File: astropy/units/core.py
Comment (nstarman) [reply astropy/units/core.py]: ✅

File: astropy/units/core.py
Comment (nstarman) [reply astropy/units/core.py]: ✅

**Comment by nstarman** [comment]
After opening https://github.com/charliermarsh/ruff/issues/1858, it was pointed out that D415 permits exclamation marks and question marks where D400 does not. I'm changing this PR to fix D415, not D400. We'll ignore D400 permanently, I guess.

File: astropy/utils/data.py
Comment (mhvk) [reply astropy/utils/data.py:172]: Sounds good.

**Comment by mhvk** [comment]
Yes, sounds good to allow all. Maybe rename this PR too, then!

File: astropy/coordinates/angles.py
Comment (mhvk) [inline astropy/coordinates/angles.py:176]: This extra `.` is wrong... I would just remove "This is a read-only property"

File: astropy/coordinates/angles.py
Comment (mhvk) [inline astropy/coordinates/angles.py:184]: Same

File: astropy/coordinates/angles.py
Comment (mhvk) [inline astropy/coordinates/angles.py:193]: Same

File: astropy/coordinates/jparser.py
Comment (mhvk) [inline astropy/coordinates/jparser.py:61]: This is wrong. Maybe because it has no description?

File: astropy/coordinates/sky_coordinate.py
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:723]: Here, instead `."` -> `".`

File: astropy/coordinates/sky_coordinate.py
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:954]: With a full stop here we probably want a `;` on the line above.

File: astropy/cosmology/core.py
Comment (mhvk) [inline astropy/cosmology/core.py:315]: Weird, why did this one escape?

File: astropy/cosmology/flrw/base.py
Comment (mhvk) [inline astropy/cosmology/flrw/base.py:335]: And this one!?

File: astropy/io/ascii/daophot.py
Comment (mhvk) [inline astropy/io/ascii/daophot.py:46]: This is wrong

File: astropy/io/fits/convenience.py
Comment (mhvk) [inline astropy/io/fits/convenience.py:5]: Huhh??

File: astropy/io/misc/yaml.py
Comment (mhvk) [inline astropy/io/misc/yaml.py:20]: I would do a line break at l6 instead; the period here is wrong.

File: astropy/io/votable/tests/vo_test.py
Comment (mhvk) [inline astropy/io/votable/tests/vo_test.py:841]: Remove period after `object`

File: astropy/io/votable/tree.py
Comment (mhvk) [inline astropy/io/votable/tree.py:1170]: Oops

File: astropy/io/votable/tree.py
Comment (mhvk) [inline astropy/io/votable/tree.py:3452]: Maybe better to split at start of `(` -- and maybe remove parentheses?

File: astropy/modeling/bounding_box.py
Comment (mhvk) [inline astropy/modeling/bounding_box.py:497]: These and the following ones do not seem very good. Should it be a `::` in the line above?

File: astropy/modeling/core.py
Comment (mhvk) [inline astropy/modeling/core.py:1025]: Here, I'd just remove the itemization.

File: astropy/modeling/core.py
Comment (mhvk) [inline astropy/modeling/core.py:1089]: Instead empty line an no indent?

File: astropy/modeling/rotations.py
Comment (mhvk) [inline astropy/modeling/rotations.py:5]: Move reference to the sentence below? (and full stop after `II`)

File: astropy/nddata/_testing.py
Comment (mhvk) [inline astropy/nddata/_testing.py:2]: Oops!

File: astropy/nddata/nduncertainty.py
Comment (mhvk) [inline astropy/nddata/nduncertainty.py:114]: Oops!

File: astropy/stats/bayesian_blocks.py
Comment (mhvk) [inline astropy/stats/bayesian_blocks.py:5]: Oops!

Comment (mhvk) [inline astropy/stats/bayesian_blocks.py:5]: Oops!

File: astropy/stats/bayesian_blocks.py
Comment (mhvk) [inline astropy/stats/bayesian_blocks.py:288]: Needs a full stop here (or semicolon and lower case `see`)

File: astropy/tests/command.py
Comment (mhvk) [inline astropy/tests/command.py:211]: Oops!

File: astropy/timeseries/periodograms/bls/__init__.py
Comment (mhvk) [inline astropy/timeseries/periodograms/bls/__init__.py:5]: Oops!

File: astropy/utils/data.py
Comment (mhvk) [inline astropy/utils/data.py:172]: Still bad?!?!

File: astropy/wcs/__init__.py
Comment (mhvk) [inline astropy/wcs/__init__.py:6]: Again a bad one. Maybe make a title sentence as you did elsewhere...

**Review by mhvk — COMMENTED** [review]
OK, I went over all of the changes. Quite a few bad ones still, but mostly this is for the better. Once you have made the changes, perhaps we can just merge; I don't think it is worth having many people go over so many...

**Comment by dhomeier** [comment]
Why has a title line (which pretty often is not a complete grammatical sentence) to end in a punctuation mark at all? If there is a PEP about this, I guess it should be done, but still seems weird writing style to me.

**Comment by nstarman** [comment]
> Are "D400" and "D415" documented anywhere?

The official designation is at http://www.pydocstyle.org/en/stable/error_codes.html but ruff also has a full listing.

**Comment by dhomeier** [comment]
> The official designation is at http://www.pydocstyle.org/en/stable/error_codes.html but ruff also has a full listing.

Thanks! Quite ironically, I failed to find any documentation for ruff at all, not expecting everything in a single README file. 😝

**Comment by nstarman** [comment]
> Thanks! Quite ironically, I failed to find any documentation for ruff at all, not expecting everything in a single README file. 😝

Haha. yeah. It's a huge README. Probably due for some docs soon.

File: astropy/coordinates/angles.py
Comment (nstarman) [reply astropy/coordinates/angles.py]: ✅. I've removed "This is a read-only property"

File: astropy/coordinates/angles.py
Comment (nstarman) [reply astropy/coordinates/angles.py]: remove "This is a read-only property"

File: astropy/coordinates/angles.py
Comment (nstarman) [reply astropy/coordinates/angles.py]: ✅. I've removed "This is a read-only property" and also reformatted so there's a summary line.

File: astropy/coordinates/jparser.py
Comment (nstarman) [reply astropy/coordinates/jparser.py:64]: Yeh, the spacing was off and there was no summary.

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py]: I added a summary line.

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py]: I added a summary line.

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py]: ✅

File: astropy/cosmology/flrw/base.py
Comment (nstarman) [reply astropy/cosmology/flrw/base.py]: ✅

File: astropy/io/ascii/daophot.py
Comment (nstarman) [reply astropy/io/ascii/daophot.py]: I added a summary line.

File: astropy/io/fits/convenience.py
Comment (nstarman) [reply astropy/io/fits/convenience.py]: I added a summary line.

File: astropy/io/misc/yaml.py
Comment (nstarman) [reply astropy/io/misc/yaml.py]: I added a summary line.

File: astropy/io/votable/tests/vo_test.py
Comment (nstarman) [reply astropy/io/votable/tests/vo_test.py]: ✅

File: astropy/io/votable/tree.py
Comment (nstarman) [reply astropy/io/votable/tree.py:1170]: ✅

File: astropy/io/votable/tree.py
Comment (nstarman) [reply astropy/io/votable/tree.py]: ✅ I added a summary line.

File: astropy/modeling/bounding_box.py
Comment (nstarman) [reply astropy/modeling/bounding_box.py]: I compressed this to a summary line.

File: astropy/modeling/core.py
Comment (nstarman) [reply astropy/modeling/core.py]: ✅

File: astropy/modeling/core.py
Comment (nstarman) [reply astropy/modeling/core.py]: ✅

File: astropy/modeling/rotations.py
Comment (nstarman) [reply astropy/modeling/rotations.py]: ✅ It actually fits well on 1 line.

File: astropy/nddata/_testing.py
Comment (nstarman) [reply astropy/nddata/_testing.py]: ✅

File: astropy/nddata/nduncertainty.py
Comment (nstarman) [reply astropy/nddata/nduncertainty.py]: ✅

File: astropy/stats/bayesian_blocks.py
Comment (nstarman) [reply astropy/stats/bayesian_blocks.py]: ✅ I added a summary line.

Comment (nstarman) [reply astropy/stats/bayesian_blocks.py]: ✅ I added a summary line.

File: astropy/stats/bayesian_blocks.py
Comment (nstarman) [reply astropy/stats/bayesian_blocks.py]: ✅

File: astropy/tests/command.py
Comment (nstarman) [reply astropy/tests/command.py]: ✅

File: astropy/timeseries/periodograms/bls/__init__.py
Comment (nstarman) [reply astropy/timeseries/periodograms/bls/__init__.py]: ✅

File: astropy/utils/data.py
Comment (nstarman) [reply astropy/utils/data.py]: ✅✅

File: astropy/wcs/__init__.py
Comment (nstarman) [reply astropy/wcs/__init__.py]: ✅

**Comment by nstarman** [comment]
Thanks @mhvk for the detailed review. I think I've worked through all the comments. I will squash the commits to 1, but kept it separate for now for review.

File: astropy/coordinates/jparser.py
Comment (mhvk) [reply astropy/coordinates/jparser.py:64]: There still is a full stop at the end of this line...

File: astropy/coordinates/sky_coordinate.py
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:724]: I'd swap the closing " with the period, i.e., "... transformation". (I'm very fond of putting the closing quotes on the outside, but it makes sense most if properly typeset, since then the period is just underneath the quotes.)

**Review by mhvk — COMMENTED** [review]
Two tiny things left!

File: astropy/wcs/__init__.py
Comment (mcara) [inline astropy/wcs/__init__.py:2]: Interesting... Where did this line come from?

File: astropy/wcs/tests/helper.py
Comment (mcara) [inline astropy/wcs/tests/helper.py:17]: I think a colon is more appropriate here. This is a string comment and ending it with a dot is not correct, IMO.

Comment (mcara) [reply astropy/wcs/tests/helper.py:17]:
```suggestion
        # set essential parameters of the model (coord transformations):
```

File: astropy/wcs/__init__.py
Comment (nstarman) [reply astropy/wcs/__init__.py:2]: For a few it was necessary to add the summary line, which was missing (https://numpydoc.readthedocs.io/en/latest/format.html#short-summary).
This occurred in a few locations of very old docs, I think before the adoption of numpydoc as our standard. This line for example was last edited in 2011.

File: astropy/coordinates/jparser.py
Comment (nstarman) [reply astropy/coordinates/jparser.py]: Fixed!

File: astropy/wcs/tests/helper.py
Comment (nstarman) [reply astropy/wcs/tests/helper.py]: ✅. I turned the docstring into a comment.

**Review by mcara — APPROVED** [review]
Looks OK for WCS. 🤞

**Review by mhvk — APPROVED** [review]
Happy with it all! As I went through all of them, I think it is OK to merge this after perhaps just another day to see if anyone else wants to have a look. Thanks!

File: pyproject.toml
Comment (eerovaher) [inline pyproject.toml:274]: The `*_lextab.py` and `*_parsetab.py` files are automatically generated, so I would exclude them from all code style checks, not just D415.

Comment (nstarman) [reply pyproject.toml]: ✅  I've added them to the exclusion list, removing them from the per-file ignores. Ruff will not run on lext/parsetab at all.

**Comment by nstarman** [comment]
> Is it possible to add comments to `pyproject.toml` explaining the new D415 exceptions?

Add comments to the per-file ignores? 

We have a comment on the D400 ignore: ""D400",  # EndsInPeriod. Using D415 instead."

**Comment by nstarman** [comment]
Added the comment "# The docstrings are regexes." to the per-file ignores for "astropy/coordinates/angle_formats.py" and "astropy/units/format/*.py". Now all have comments explaining the ignore.

**Review by eerovaher — COMMENTED** [review]
All my suggestions about the final state of the code have been implemented already, but I think telling `ruff` to ignore `astropy/extern` and the automatically generated `*_parsetab.py` and `*_lextab.py` files could be done in a separate commit.

**Comment by nstarman** [comment]
> All my suggestions about the final state of the code have been implemented already, but I think telling `ruff` to ignore `astropy/extern` and the automatically generated `*_parsetab.py` and `*_lextab.py` files could be done in a separate commit.

It actually is done in a previous commit. This PR just moved the config up and prepended "_" to the parrsetab, lextab exclusion, which is better but actually makes no difference matching to those files.

File: pyproject.toml
Comment (eerovaher) [inline pyproject.toml:82]: Was there some good reason for moving these lines around? If not then I'd recommend leaving the `exclude` where and as it was.

Comment (nstarman) [reply pyproject.toml:82]: A few, of varying qualities of good.
1. alphabetical.
2. I wanted ``extend-ignore`` closer to the ``per-file ignores`` as both are for check-specific ignores, while ``exclude`` ignore all checks.

**Review by eerovaher — APPROVED** [review]
I think the `exclude` list in `pyproject.toml` could have stayed where it was, but that shouldn't block this pull request from being merged.

**Pull request merged**
